File: app/src/test/java/com/elegion/android/LifecyclerTest.java
package com.elegion.android;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.Fragment;
import android.app.FragmentManager;
import android.app.FragmentTransaction;
import android.os.Bundle;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.Spy;
import org.mockito.runners.MockitoJUnitRunner;

import static org.hamcrest.core.Is.is;
import static org.junit.Assert.assertThat;

/**
 * @author Daniel Serdyukov
 */
@SuppressLint("CommitTransaction")
@RunWith(MockitoJUnitRunner.class)
public class LifecyclerTest {

    @Mock
    private Activity mActivity;

    @Mock
    private Bundle mSavedInstanceState;

    @Mock
    private FragmentManager mFm;

    @Mock
    private FragmentTransaction mFt;

    @Spy
    private Lifecycler mLifecycler = new Lifecycler();

    @Before
    public void setUp() throws Exception {
        Mockito.doReturn(mFm).when(mActivity).getFragmentManager();
        Mockito.doReturn(mFt).when(mFm).beginTransaction();
        Mockito.doReturn(mFt).when(mFt).add(Mockito.<Fragment>any(), Mockito.anyString());
    }

    @Test
    public void testOnActivityStarted() throws Exception {
        mLifecycler.onActivityStarted(mActivity);
        Mockito.verifyZeroInteractions(mActivity);
    }

    @Test
    public void testOnActivityResumed() throws Exception {
        mLifecycler.onActivityResumed(mActivity);
        Mockito.verifyZeroInteractions(mActivity);
    }

    @Test
    public void testOnActivityPaused() throws Exception {
        mLifecycler.onActivityPaused(mActivity);
        Mockito.verifyZeroInteractions(mActivity);
    }

    @Test
    public void testOnActivityStopped() throws Exception {
        mLifecycler.onActivityStopped(mActivity);
        Mockito.verifyZeroInteractions(mActivity);
    }

    @Test
    public void testOnActivitySaveInstanceState() throws Exception {
        mLifecycler.onActivitySaveInstanceState(mActivity, mSavedInstanceState);
        Mockito.verifyZeroInteractions(mActivity);
        Mockito.verifyZeroInteractions(mSavedInstanceState);
    }

    @Test
    public void testOnActivityDestroyed() throws Exception {
        mLifecycler.onActivityDestroyed(mActivity);
        Mockito.verifyZeroInteractions(mActivity);
    }

    @Test
    public void testIsApplicationVisible() throws Exception {
        assertThat(Lifecycler.isApplicationVisible(), is(false));
        mLifecycler.onActivityStarted(mActivity);
        assertThat(Lifecycler.isApplicationVisible(), is(true));
        mLifecycler.onActivityStopped(mActivity);
        assertThat(Lifecycler.isApplicationVisible(), is(false));
    }

    @Test
    public void testOnActivityCreated() throws Exception {
        mLifecycler.onActivityCreated(mActivity, mSavedInstanceState);
        Mockito.verifyZeroInteractions(mActivity);
    }
}

File: app/src/test/kotlin/com/elegion/android/template/LifecyclerTest.kt
package com.elegion.android.template

import android.annotation.SuppressLint
import android.app.Activity
import android.app.FragmentManager
import android.app.FragmentTransaction
import android.os.Bundle
import org.hamcrest.core.Is.`is`
import org.junit.Assert.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.Spy
import org.mockito.junit.MockitoJUnitRunner

@SuppressLint("CommitTransaction")
@RunWith(MockitoJUnitRunner.Silent::class)
class LifecyclerTest {

    @Mock
    private val activity: Activity? = null

    @Mock
    private val savedInstanceState: Bundle? = null

    @Mock
    private val fragmentManager: FragmentManager? = null

    @Mock
    private val fragmentTransaction: FragmentTransaction? = null

    @Spy
    private val lifecycler = Lifecycler()

    @Before
    @Throws(Exception::class)
    fun setUp() {
        Mockito.doReturn(fragmentManager).`when`<Activity>(activity).fragmentManager
        Mockito.doReturn(fragmentTransaction).`when`<FragmentManager>(fragmentManager).beginTransaction()
        Mockito.doReturn(fragmentTransaction).`when`<FragmentTransaction>(fragmentTransaction).add(Mockito.any(), Mockito.anyString())
    }

    @Test
    @Throws(Exception::class)
    fun testOnActivityStarted() {
        lifecycler.onActivityStarted(activity!!)
        Mockito.verifyZeroInteractions(activity)
    }

    @Test
    @Throws(Exception::class)
    fun testOnActivityResumed() {
        lifecycler.onActivityResumed(activity!!)
        Mockito.verifyZeroInteractions(activity)
    }

    @Test
    @Throws(Exception::class)
    fun testOnActivityPaused() {
        lifecycler.onActivityPaused(activity!!)
        Mockito.verifyZeroInteractions(activity)
    }

    @Test
    @Throws(Exception::class)
    fun testOnActivityStopped() {
        lifecycler.onActivityStopped(activity!!)
        Mockito.verifyZeroInteractions(activity)
    }

    @Test
    @Throws(Exception::class)
    fun testOnActivitySaveInstanceState() {
        lifecycler.onActivitySaveInstanceState(activity!!, savedInstanceState!!)
        Mockito.verifyZeroInteractions(activity)
        Mockito.verifyZeroInteractions(savedInstanceState)
    }

    @Test
    @Throws(Exception::class)
    fun testOnActivityDestroyed() {
        lifecycler.onActivityDestroyed(activity!!)
        Mockito.verifyZeroInteractions(activity)
    }

    @Test
    @Throws(Exception::class)
    fun testIsApplicationVisible() {
        assertThat(Lifecycler.isApplicationVisible, `is`(false))
        lifecycler.onActivityStarted(activity!!)
        assertThat(Lifecycler.isApplicationVisible, `is`(true))
        lifecycler.onActivityStopped(activity)
        assertThat(Lifecycler.isApplicationVisible, `is`(false))
    }

    @Test
    @Throws(Exception::class)
    fun testOnActivityCreated() {
        lifecycler.onActivityCreated(activity!!, savedInstanceState!!)
        Mockito.verifyZeroInteractions(activity)
    }
}